src/kernels/jupyter/finder/universalRemoteKernelFinder.ts

```diff
@@ -0,0 +1,397 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { CancellationToken, CancellationTokenSource, Event, EventEmitter, Memento, Uri } from 'vscode';
+import { getKernelId, getLanguageInKernelSpec, serializeKernelConnection } from '../../helpers';
+import {
+    IJupyterKernelSpec,
+    IKernelProvider,
+    INotebookProvider,
+    INotebookProviderConnection,
+    isRemoteConnection,
+    KernelConnectionMetadata,
+    LiveRemoteKernelConnectionMetadata,
+    RemoteKernelConnectionMetadata,
+    RemoteKernelSpecConnectionMetadata
+} from '../../types';
+import { IDisposable, IExtensions, Resource } from '../../../platform/common/types';
+import { IInterpreterService } from '../../../platform/interpreter/contracts';
+import { capturePerfTelemetry, Telemetry } from '../../../telemetry';
+import {
+    IJupyterSessionManagerFactory,
+    IJupyterSessionManager,
+    IJupyterRemoteCachedKernelValidator,
+    IRemoteKernelFinder,
+    IJupyterServerUriEntry
+} from '../types';
+import { sendKernelSpecTelemetry } from '../../raw/finder/helper';
+import { traceError, traceWarning, traceInfoIfCI } from '../../../platform/logging';
+import { IPythonExtensionChecker } from '../../../platform/api/types';
+import { computeServerId } from '../jupyterUtils';
+import { PYTHON_LANGUAGE } from '../../../platform/common/constants';
+import { createPromiseFromCancellation } from '../../../platform/common/cancellation';
+import { DisplayOptions } from '../../displayOptions';
+import * as localize from '../../../platform/common/utils/localize';
+import { isArray } from '../../../platform/common/utils/sysTypes';
+import { deserializeKernelConnection } from '../../helpers';
+import { noop } from '../../../platform/common/utils/misc';
+import { IApplicationEnvironment } from '../../../platform/common/application/types';
+import { KernelFinder } from '../../kernelFinder';
+import { RemoteKernelSpecsCacheKey, removeOldCachedItems } from '../../common/commonFinder';
+import { ContributedKernelFinderKind, IContributedKernelFinderInfo } from '../../internalTypes';
+import { disposeAllDisposables } from '../../../platform/common/helpers';
+
+// Even after shutting down a kernel, the server API still returns the old information.
+// Re-query after 2 seconds to ensure we don't get stale information.
+const REMOTE_KERNEL_REFRESH_INTERVAL = 2_000;
+
+// This class watches a single jupyter server URI and returns kernels from it
+export class UniversalRemoteKernelFinder implements IRemoteKernelFinder, IContributedKernelFinderInfo, IDisposable {
```

Comment:
curious why the name universal?
if the old class didn't exist would we still call this universal or just remotekernel finder?

Reply:
Yeah, it was just a way to distinguish between the existing remote kernel finder and the prototype finder. Universal being universally considering all servers.